### DISCUSSION TITLE
Build help

### DISCUSSION THREAD

**Discussion by ruzam18** [cockpit-project/cockpit-files#583] [General]
Hi there - 
I'm trying to build this on Debian arm64, running into this:

"Specifically the "@esbuild/linux-x64" package is present but this platform
needs the "@esbuild/linux-arm64" package instead."

I notice there was a similar issue on fedora: [https://github.com/cockpit-project/cockpit-files/commit/84d8dc5db00a490b1d656177fd25ec26d1396e79](url)

Wondered if someone could help me figure this out? Thanks!

**Comment by martinpitt** [comment]
This is an awkward npm check indeed. There's already esbuild-wasm which `./build.js` will use, but that silly check breaks it. As a workaround, please run `tools/node-modules install`. If that doesn't work for some reason, you can also run `npm install`.

**Comment by ruzam18** [comment]
Hi Martin,

I've run that, but it doesn't work. That errors out too.

Kind Regards,
Samuel

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by martinpitt** [reply]
Try:
```
rm -rf node_modules
npm install
```

**Comment by ruzam18** [comment]
I'm not a developer, I'm a fiddler! I normally use Fedora, but have to use Debian for this.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ruzam18** [comment]
I've done that too.
It hangs up again on a non arm64 library.
I've looked through the code but can't find the x64 reference for Debian.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by martinpitt** [reply]
Sorry then, I can't think of anything else. packages.json has *nothing* specific to x86_64 -- If a mere `npm install esbuild` doesn't work on ARM, that's a bug for https://github.com/evanw/esbuild/issues

**Comment by ruzam18** [comment]
So I do npm install, make. NPM works, make is where it dies.
Can I build the artifacts without make?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ruzam18** [comment]
Yeah, I checked that too.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ruzam18** [comment]
Rephrase. Esbuild with Cockpit is selecting the x86 library on arm64 when the arm library is present.
Samuel

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ruzam18** [comment]
I'll look into it tomorrow, . Thankyou Martin. Apologies for my non-developer angle.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tidexe** [comment]
same question 😭

**Reply by martinpitt** [reply]
`rm -rf node_modules; npm install` was reported to work in other places. As I said above, if that doesn't work, I have no idea at all what's wrong, but it can't possibly be the same error message as in the description. What does the error look like then?

A different approach is to just build the project on an x86_64 machine, run `make install DESTDIR=/tmp/i` and copy /tmp/i/ to the target machine. There  are no compiled binaries there, just data.

**Reply by tidexe** [reply]
It works!!!
Thank you martin.😆
I copy the data from my x86 machine,then put the files into another arm64 machine.

**Comment by begunfx** [comment]
Same problem here.  Install works make does not.  I get the same error at the OP.

```
make package-lock.json && NODE_ENV= ./build.js
make[1]: Entering directory '/usr/local/share/cockpit-files'
tools/node-modules make_package_lock_json
make[1]: Leaving directory '/usr/local/share/cockpit-files'

node:internal/modules/run_main:105
    triggerUncaughtException(
    ^
Error [TransformError]: 
You installed esbuild for another platform than the one you're currently using.
This won't work because esbuild is written with native code and needs to
install a platform-specific binary executable.

Specifically the "@esbuild/linux-x64" package is present but this platform
needs the "@esbuild/linux-arm64" package instead. People often get into this
situation by installing esbuild on Windows or macOS and copying "node_modules"
into a Docker image that runs Linux, or by copying "node_modules" between
Windows and WSL environments.

If you are installing with npm, you can try not copying the "node_modules"
directory when you copy the files over, and running "npm ci" or "npm install"
on the destination platform after the copy. Or you could consider using yarn
instead of npm which has built-in support for installing a package on multiple
platforms simultaneously.

If you are installing with yarn, you can try listing both this platform and the
other platform in your ".yarnrc.yml" file using the "supportedArchitectures"
feature: https://yarnpkg.com/configuration/yarnrc/#supportedArchitectures
Keep in mind that this means multiple copies of esbuild will be present.

Another alternative is to use the "esbuild-wasm" package instead, which works
the same way on all platforms. But it comes with a heavy performance cost and
can sometimes be 10x slower than the "esbuild" package, so you may also not
want to do that.

    at generateBinPath (/usr/local/share/cockpit-files/node_modules/esbuild/lib/main.js:1734:17)
    at esbuildCommandAndArgs (/usr/local/share/cockpit-files/node_modules/esbuild/lib/main.js:1815:33)
    at ensureServiceIsRunning (/usr/local/share/cockpit-files/node_modules/esbuild/lib/main.js:1972:25)
    at transform (/usr/local/share/cockpit-files/node_modules/esbuild/lib/main.js:1873:37)
    at file:///usr/local/share/cockpit-files/node_modules/tsx/dist/index-DGv_vkxZ.mjs:14:2755
    at applyTransformers (file:///usr/local/share/cockpit-files/node_modules/tsx/dist/index-DGv_vkxZ.mjs:14:1266)
    at transform (file:///usr/local/share/cockpit-files/node_modules/tsx/dist/index-DGv_vkxZ.mjs:14:2702)
    at load (file:///usr/local/share/cockpit-files/node_modules/tsx/dist/esm/index.mjs?1748563315456:2:2245)
    at async nextLoad (node:internal/modules/esm/hooks:748:22)
    at async Hooks.load (node:internal/modules/esm/hooks:385:20)

Node.js v24.1.0
make: *** [Makefile:92: dist/manifest.json] Error 1
```

**Comment by begunfx** [comment]
I think I finally solved it.

The following worked. 

```
rm -fr node_modules
npm install
./build.js
```

using `./build.js` instead of `make` seemed to work on Raspian OS (Debian)
